metron-deployment/vagrant/KERBEROS_SETUP.md

```diff
@@ -0,0 +1,238 @@
+```
+
+14. Add metron user to the Kafka cluster ACL
+```
+/usr/hdp/current/kafka-broker/bin/kafka-acls.sh --authorizer kafka.security.auth.SimpleAclAuthorizer --authorizer-properties zookeeper.connect=node1:2181 --add --allow-principal User:${KERB_USER} --cluster kafka-cluster
+```
+
+15. We also need to grant permissions to the HBase tables. Kinit as the hbase user and add ACLs for metron.
+```
+kinit -kt /etc/security/keytabs/hbase.headless.keytab hbase-metron_cluster@EXAMPLE.COM
+echo "grant 'metron', 'RW', 'threatintel'" | hbase shell
+echo "grant 'metron', 'RW', enrichment" | hbase shell
+```
+
+16. Create a “.storm” directory in the metron user’s home directory and switch to that directory.
+```
+su - metron
+mkdir .storm
+cd .storm
+```
+
+17. Create a custom client jaas file. This should look identical to the Storm client jaas file located in /etc/storm/conf/client_jaas.conf except for the addition of a Client stanza. The Client stanza is used for Zookeeper. All quotes and semicolons are necessary.
+```
+[metron@node1 .storm]$ cat client_jaas.conf
+StormClient {
+   com.sun.security.auth.module.Krb5LoginModule required
+   useTicketCache=true
+   renewTicket=true
+   serviceName="nimbus";
+};
+Client {
+   com.sun.security.auth.module.Krb5LoginModule required
+   useTicketCache=true
+   renewTicket=true
+   serviceName="zookeeper";
+};
+KafkaClient {
+   com.sun.security.auth.module.Krb5LoginModule required
+   useTicketCache=true
+   renewTicket=true
+   serviceName="kafka";
+};
+```
+
+18. Create a storm.yaml with jaas file info.
+```
+[metron@node1 .storm]$ cat storm.yaml
+nimbus.seeds : ['node1']
+java.security.auth.login.config : '/home/metron/.storm/client_jaas.conf'
+storm.thrift.transport : 'org.apache.storm.security.auth.kerberos.KerberosSaslTransportPlugin'
+```
+
+19. Create an auxiliary storm configuration json file in the metron user’s home directory. Note the login config option in the file points to our custom client_jaas.conf.
+```
+cd /home/metron
+[metron@node1 ~]$ cat storm-config.json
+{
+  "topology.worker.childopts" : "-Djava.security.auth.login.config=/home/metron/.storm/client_jaas.conf"
+}
+```
+
+20. Setup enrichment and indexing.
+    1. Modify enrichment.properties - `${METRON_HOME}/config/enrichment.properties`
+```
+kafka.security.protocol=PLAINTEXTSASL
+topology.worker.childopts=-Djava.security.auth.login.config=/home/metron/.storm/client_jaas.conf
+```
+
+    2. Modify elasticsearch.properties - `${METRON_HOME}/config/elasticsearch.properties`
+```
+kafka.security.protocol=PLAINTEXTSASL
+topology.worker.childopts=-Djava.security.auth.login.config=/home/metron/.storm/client_jaas.conf
+```
+
+21. Restart the parser topologies. Be sure to pass in the new parameter, “-ksp” or “--kafka_security_protocol.” Run this from the metron home directory.
+```
+for parser in bro snort; do ${METRON_HOME}/bin/start_parser_topology.sh -z node1:2181 -s ${parser} -ksp PLAINTEXTSASL -e storm-config.json; done
```

Comment:
At this step however, I am getting the following error. The full log message with the command is available [here](https://pastebin.com/ynHzkeWy). 

Any idea why I am seeing this error?

`
3770 [main] WARN  o.a.s.s.a.k.ClientCallbackHandler - Could not login: the client is being asked for a password, but the  client code does not currently support obtaining a password from the user. Make sure that the client is configured to use a ticket cache (using the JAAS configuration setting 'useTicketCache=true)' and restart the client. If you still get this message after that, the TGT in the ticket cache has expired and must be manually refreshed. To do so, first determine if you are using a password or a keytab. If the former, run kinit in a Unix shell in the environment of the user who is running this client using the command 'kinit <princ>' (where <princ> is the name of the client's Kerberos principal). If the latter, do 'kinit -k -t <keytab> <princ>' (where <princ> is the name of the Kerberos principal, and <keytab> is the location of the keytab file). After manually refreshing your cache, restart this client. If you continue to see this message after manually refreshing your cache, ensure that your KDC host's clock is in sync with this host's clock.
java.lang.RuntimeException: javax.security.auth.login.LoginException: No password provided
	at org.apache.storm.security.auth.kerberos.AutoTGT.populateCredentials(AutoTGT.java:103)
	at org.apache.storm.StormSubmitter.populateCredentials(StormSubmitter.java:94)
	at org.apache.storm.StormSubmitter.submitTopologyAs(StormSubmitter.java:214)
	at org.apache.storm.StormSubmitter.submitTopology(StormSubmitter.java:310)
	at org.apache.storm.StormSubmitter.submitTopology(StormSubmitter.java:157)
	at org.apache.metron.parsers.topology.ParserTopologyCLI.main(ParserTopologyCLI.java:318)
Caused by: javax.security.auth.login.LoginException: No password provided
	at com.sun.security.auth.module.Krb5LoginModule.promptForPass(Krb5LoginModule.java:919)
	at com.sun.security.auth.module.Krb5LoginModule.attemptAuthentication(Krb5LoginModule.java:760)
	at com.sun.security.auth.module.Krb5LoginModule.login(Krb5LoginModule.java:617)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at javax.security.auth.login.LoginContext.invoke(LoginContext.java:755)
	at javax.security.auth.login.LoginContext.access$000(LoginContext.java:195)
	at javax.security.auth.login.LoginContext$4.run(LoginContext.java:682)
	at javax.security.auth.login.LoginContext$4.run(LoginContext.java:680)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:680)
	at javax.security.auth.login.LoginContext.login(LoginContext.java:587)
	at org.apache.storm.security.auth.kerberos.AutoTGT.populateCredentials(AutoTGT.java:80)
	... 5 more
`